### PR TITLE
Implement remote control support

### PR DESCRIPTION
**Pull request by hristoterezov**
The PR is still WIP. Currently implemented:

- capturing keyboard and mouse events on the controller side

- sending captured events to the receiver side

- listening for controller events on the receiver side and sending them trough the iframe API to external application that will execute them on the receiver side.

- Announce remotecontrol support from the receiver side to all participants

- Implements request for remote control permissions.

Depending on:
https://github.com/jitsi/lib-jitsi-meet/pull/341

### CODE REVIEW

**Comment by paweldomas** [comment]
jenkins test this please

**Comment by paweldomas** [comment]
the tests are failing on compilation with:
15:16:12     ERROR in ./app.js
15:16:12     Module not found: Error: Cannot resolve 'file' or 'directory' ./modules/remotecontrol/remotecontrol in /home/jenkins/jitsi-meet

File: conference.js [hunk 1]
```diff
@@ -1800,5 +1818,17 @@ export default {
             APP.API.notifyReadyToClose();
             maybeRedirectToWelcomePage(values[0]);
         });
+    },
+
+    /**
+     * Sends a message via the data channel.
+     * @param to {string} the id of the endpoint that should receive the
```
Comment (paweldomas) [inline conference.js]: First goes the type then goes the name

File: modules/API/API.js [hunk 2]
```diff
@@ -257,6 +257,14 @@ export default {
     },
 
     /**
+     * Sends remote control event.
```
Comment (paweldomas) [inline modules/API/API.js]: Can you provide more details about this method and it's arguments ? What structure does the event have or where does it's value come from ?

File: modules/keycode/keycode.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    PERIOD: ".",
+    SEMICOLON: ";",
+    QUOTE: "'",
+    BRACKET_LEFT: "[",
+    BRACKET_RIGHT: "]",
+    BACKQUOTE: "`",
+    BACKSLASH: "\\",
+    MINUS: "-",
+    EQUAL: "=",
+    SLASH: "/"
```
Comment (paweldomas) [inline modules/keycode/keycode.js]: How about " ~ ! @ # $ % ^ & * ( ) + ? Are those not supported ?

File: modules/keycode/keycode.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+/**
+ * Generate codes for digit keys (0-9)
+ */
+for(let i = 0; i < 10; i++) {
+    keyCodeToKey[i + 48] = `${i}`;
+}
+
+/**
+ * Generate codes for letter keys (a-z)
```
Comment (paweldomas) [inline modules/keycode/keycode.js]: I wonder if it would be better to capture output character that the keyboard produces with current user's layout instead of doing mapping. How different languages are going to be handled here ? What will happen with non-latin alphabet letters ? Are they going to work ?

File: modules/remotecontrol/Controller.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            this._onMouseClickHandler.bind(this, "mousedblclick"));
+        this.area.contextmenu(() => false);
+        this.area[0].onmousewheel = event => {
+            this._sendEvent({
+                type: "mousescroll",
+                x: event.deltaX,
+                y: event.deltaY
+            });
+        };
+        $(window).keydown(this._onKeyPessHandler.bind(this, "keydown"));
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: I wonder if the keypress event should be handled as well. Normally when you long press the key it will be printing characters repeatedly.

File: modules/remotecontrol/Controller.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    /**
+     * Starts processing the mouse and keyboard events.
+     * @param {JQuery.selector} area the selector which will be used for
+     * attaching the listeners on.
+     */
+    start(area) {
+        this.area = area;
+        this.area.mousemove(event => {
+            const position = this.area.position();
+            this._sendEvent({
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Where is "_sendEvent" method declared ?

File: modules/remotecontrol/Controller.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        });
+    }
+
+    /**
+     * Sends remote control event to the controlled participant.
+     * @param {Object} event the remote control event.
+     */
+    _sendRemoteControlEvent(event) {
+        try{
+            APP.conference.sendEndpointMessage("",
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Why is the message broadcasted instead of being sent directly to the controlled endpoint? A controller controls only one endpoint at a time, right ?

File: modules/remotecontrol/Controller.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                {type: "remote-control-event", event});
+        } catch (e) {
+            // failed to send the event.
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Why errors are ignored ?

File: modules/remotecontrol/Receiver.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    stop() {
+        APP.conference.removeConferenceListener(
+            ConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            this._onRemoteControlEvent);
+    }
+
+    /**
+     * Sends "remote-control-event" events to to the API module.
+     * @param {JitsiParticipant} participant the controller participant
+     * @param {Object} event the remote control event.
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: "event" is not "the remote control event", but it's a structure received over a DataChannel. event.event is expected to be the remote control event, so the comment is slightly misleading.

File: modules/remotecontrol/Receiver.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
+/* global APP, JitsiMeetJS */
+const ConferenceEvents = JitsiMeetJS.events.conference;
+
+/**
+ * This class represents the receiver party for a remote controller session.
+ * It handles "remote-control-event" events and sends them to the
+ * API module. From there the events can be received from wrapper application
+ * and executed.
+ */
+class Receiver {
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: I would say that a Receiver should be created for a specific endpoint for which the user has granted the permission and accept events only coming from that endpoint.

File: app.js [hunk 11]
```diff
@@ -301,9 +301,11 @@ $(document).ready(function () {
 
     APP.translation.init(settings.getLanguage());
 
-    APP.API.init(APP.tokenData.externalAPISettings);
-    // For remote control testing:
-    // APP.API.init({forceEnable: true});
+    APP.API.init(APP.tokenData.jwt ? {
```
Comment (paweldomas) [inline app.js]: What is the meaning of the code here ? If there is jwt token then force enable the API and enable the events listed ? Would you mind writing a comment here on why is that necessary ? Without seeing the rest of the commit I guess that it's required for remote control, but that's only, because I know I'm in the context of remote control PR. Other readers will have no idea.

File: modules/remotecontrol/RemoteControl.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * Handles API event for support for executing remote control events into
+     * the wrapper application.
+     * @param {boolean} isSupported true if the receiver side is supported by
+     * the wrapper application.
+     */
+    onRemoteControlSupported(isSupported) {
+        if(isSupported && !config.disableRemoteControl) {
```
Comment (paweldomas) [inline modules/remotecontrol/RemoteControl.js]: It would be helpful to have some info message logged that will indicate whether or not the current client supports remote control. Also if it's supported, but disabled.

File: modules/remotecontrol/RemoteControl.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * Initializes the remote control - checks if the remote control should be
+     * enabled or not, initializes the API module.
+     */
+    init() {
+        if(config.disableRemoteControl || this.initialized) {
+            return;
+        }
+        this.initialized = true;
+        APP.API.init({
```
Comment (paweldomas) [inline modules/remotecontrol/RemoteControl.js]: This duplicates "APP.API.init()" called in the app.js. Why is that ?

File: app.js (hunk 11 above)
Comment (paweldomas) [inline app.js]: Now I see that it was moved from TokenData.js. Still some comment would be helpful. I wonder how is the change related to remote control.

File: modules/remotecontrol/Controller.js (hunk 6 above)
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Oh I see it was changed in next commit. Please avoid doing stuff like that in future as it makes the review process harder. Especially if someone is looking at each commit one by one instead of all changes.

File: modules/remotecontrol/Controller.js [hunk 14]
```diff
@@ -42,41 +44,56 @@ function getModifiers(event) {
...
+        this.enabled = enabled;
+    }
 
     /**
      * Starts processing the mouse and keyboard events.
      * @param {JQuery.selector} area the selector which will be used for
      * attaching the listeners on.
      */
     start(area) {
+        if(!this.enabled)
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Is it expected that Controller will be started if not enabled ? I would rather see here some error logged or thrown in such case instead of ignoring it silently.

File: modules/remotecontrol/Controller.js [hunk 15]
```diff
@@ -123,14 +140,13 @@ class Controller {
      * @param {Object} event the remote control event.
      */
     _sendRemoteControlEvent(event) {
+        if(!this.enabled)
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Is it expected that _sendRemoteControlEvent will be called if Controller is not enabled ? I would rather see here some error logged or thrown in such case instead of ignoring it silently.

File: modules/remotecontrol/Receiver.js [hunk 16]
```diff
@@ -7,20 +10,36 @@ const ConferenceEvents = JitsiMeetJS.events.conference;
...
+    constructor() {
+        this.enabled = false;
+    }
+
+    /**
+     * Enables / Disables the remote control
+     * @param {boolean} enabled the new state.
+     */
+    enable(enabled) {
+        if(this.enabled !== enabled && enabled === true) {
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: With this condition it's impossible to set 'enabled' to a value other than 'true'.

File: modules/remotecontrol/Receiver.js [hunk 17]
```diff
@@ -7,20 +10,36 @@ const ConferenceEvents = JitsiMeetJS.events.conference;
...
+    }
 
     /**
      * Attaches listener for ConferenceEvents.ENDPOINT_MESSAGE_RECEIVED events.
      */
     start() {
-        APP.conference.addConferenceListener(
-            ConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
-            this._onRemoteControlEvent);
+        if(this.enabled) {
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: Same as similar comments above. I find it strange to allow calling start when Receiver is not enabled without logging or throwing any errors.

File: modules/remotecontrol/Receiver.js [hunk 18]
```diff
@@ -39,9 +58,7 @@ class Receiver {
      * @param {Object} event the remote control event.
      */
     _onRemoteControlEvent(participant, event) {
-        if(event.type === "remote-control-event")
+        if(event.type === API_EVENT_TYPE && this.enabled)
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: It would be good to log some warning here for debugging purposes if the method is called when the Receiver is not enabled.

File: modules/remotecontrol/Controller.js [hunk 19]
```diff
@@ -44,34 +47,113 @@ function getModifiers(event) {
...
      */
-    start(area) {
+    _start() {
         if(!this.enabled)
             return;
-        this.area = area;
+        APP.conference.addConferenceListener(
+            ConferenceEvents.ENDPOINT_MESSAGE_RECEIVED,
+            this.stopListener);
+        this.area = $("#largeVideoWrapper");
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: I don't really like how large video wrapper selector is hardcoded here. Would it be possible to get it somehow from the video layout or pass it as an argument ?

File: modules/remotecontrol/RemoteControl.js [hunk 20]
```diff
@@ -33,13 +35,27 @@ class RemoteControl {
     }
 
     /**
+     * Handles remote control events from the API module.
+     * @param {object} event the remote control event
```
Comment (paweldomas) [inline modules/remotecontrol/RemoteControl.js]: Some description of the event structure in the docs would be helpful.

File: conference.js [hunk 21]
```diff
@@ -1799,6 +1803,7 @@ export default {
      */
     hangup (requestFeedback = false) {
         APP.UI.hideRingOverLay();
+        APP.remoteControl.receiver.enable(false);
```
Comment (paweldomas) [inline conference.js]: For me it makes more sense to have some event for hangup/conference ended and handle it in the remoteControl.receiver directly

File: modules/UI/videolayout/RemoteVideo.js [hunk 22]
```diff
@@ -182,6 +205,68 @@ RemoteVideo.prototype._generatePopupContent = function () {
     return popupmenuElement;
 };
 
+/**
+ * Sets the remote control supported value and initializes or updates the menu
+ * depending on the remote control is supported or not.
+ * @param {boolean} isSupported
+ */
+RemoteVideo.prototype.setRemoteControlSupport = function(isSupported = false) {
```
Comment (paweldomas) [inline modules/UI/videolayout/RemoteVideo.js]: It would be useful to have some info logged about participant's support for remote control (for debugging purpose).

File: modules/remotecontrol/Controller.js [hunk 23]
```diff
@@ -209,6 +253,22 @@ export default class Controller extends RemoteControlParticipant {
...
+     * Returns true if the remote control session is started.
+     * @returns {boolean}
+     */
+    isStarted() {
+        return this.controlledParticipant !== null;
+    }
+
+    /**
+     * Returns the id of the requested participant
+     * @returns {string} this.requestedParticipant
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: it would be helpful to write if the id is just an endpoint ID (MUC nickname) or full MUC JID

File: modules/UI/videolayout/RemoteVideo.js [hunk 24]
```diff
@@ -245,6 +245,10 @@ RemoteVideo.prototype._requestRemoteControlPermissions = function () {
             {user: this.user.getDisplayName()
                 || interfaceConfig.DEFAULT_REMOTE_DISPLAY_NAME}
         );
+        let pinnedId = this.VideoLayout.getPinnedId();
```
Comment (paweldomas) [inline modules/UI/videolayout/RemoteVideo.js]: Can you explain what is happening in these 4 lines ?

File: modules/keycode/keycode.js (hunk 3 above)
Comment (hristoterezov) [inline modules/keycode/keycode.js]: This map should represent the keys on the keyboard not the characters. So all these chars are going to be supported by pressing shift and number key. But I'll add more comments about it.

File: modules/keycode/keycode.js (hunk 4 above)
Comment (hristoterezov) [inline modules/keycode/keycode.js]: Currently we are doing our best to guess the key on the controller side. So it's not so much about the different languages but whether we can guess the key or not. Basically if we guess the key correctly everything will work fine(with non-latin letters and with latin letters). But there are some keys that will have conflicts and in some cases we won't be able to guess the key.

Actually currently we can't do much even if we capture the output because robotjs (the library which should execute the event on the controlled side) accepts ascii chars and than using the current keyboard type and layout produces the output there.

But even if it was possible I think if we go with this approach there are some cases that I don't know how we should handle - like what if the controlled side doesn't support this language, etc.

Also personally I think that it's less confusing if the controlled side's layout is the leading one.

File: modules/remotecontrol/Controller.js (hunk 5 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: It's working even now. No need to handle the keypress event.

File: modules/remotecontrol/Controller.js (hunk 7 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: Fixed in next commits.

File: modules/remotecontrol/Controller.js (hunk 8 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: Fixed in next commits.

File: modules/remotecontrol/Receiver.js (hunk 10 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: I like the idea. I'll thing about how to implement it. Maybe I can do the same for the controller and move the whole requesting permissions thing to the RemoteControl.js. WDYT?

File: app.js (hunk 11 above)
Comment (hristoterezov) [inline app.js]: I was making changes to API.init to be called multiple time which is needed by remote control and decided that this code doesn't belong to tokenData module. I'll add comments.

File: modules/remotecontrol/RemoteControl.js (hunk 13 above)
Comment (hristoterezov) [inline modules/remotecontrol/RemoteControl.js]: To enable the API module even if it is not enabled currently.

File: modules/remotecontrol/Controller.js (hunk 14 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: Actually we don't have such use case so I'm removing it.

File: modules/remotecontrol/Receiver.js (hunk 16 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: Fixed in next commits

File: modules/remotecontrol/Receiver.js (hunk 17 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: This one is outdated comment, right?

File: modules/remotecontrol/Controller.js (hunk 23 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: I was thinking that we are not using full JIDs in jitsi meet. Am I wrong?

File: modules/UI/videolayout/RemoteVideo.js (hunk 24 above)
Comment (hristoterezov) [inline modules/UI/videolayout/RemoteVideo.js]: Pinning the controlled participant when the reply from the request for permissions is received. Actually even if the controlled participant has denied the request, we will pin him again. Do you think we should pin the participant only if the permissions have been granted?

File: modules/API/API.js [hunk 25]
```diff
@@ -218,7 +218,18 @@ class API {
 
     /**
      * Sends remote control event.
-     * @param {object} event the event.
+     * @param {object} event the remote control event. The remote control event
```
Comment (paweldomas) [inline modules/API/API.js]: It would make more sense to move events descriptions to /service/remote/constants and describe each one there. Then just paste a reference or type here.
These links can be helpful:
http://usejsdoc.org/tags-typedef.html
http://usejsdoc.org/tags-enum.html
http://stackoverflow.com/questions/6460604/how-to-describe-object-arguments-in-jsdoc

File: modules/keycode/keycode.js [hunk 26]
```diff
@@ -1,5 +1,6 @@
 /**
  * Enumerates the supported keys.
+ * NOTE: The maps represents actual keys on the keyboard not chars.
```
Comment (paweldomas) [inline modules/keycode/keycode.js]: Is this key set related somehow to what robotjs supports ? Maybe some link to robotjs docs ?

File: modules/keycode/keycode.js (hunk 4 above)
Comment (paweldomas) [inline modules/keycode/keycode.js]: Yes, it makes sense.

File: modules/keycode/keycode.js (hunk 26 above)
Comment (hristoterezov) [inline modules/keycode/keycode.js]: What do you refer by "key"? In the "NOTE:" I mean hardware key on the keyboard device :)

Although if you are talking about the values in this case they represent the values from robotjs. But this is just enum. If we replace robotjs with something else we can continue using this enum and we'll can make mapping into the wrapper between these values and ones that the new library expect to receive. So I don't think this is a good place to add such a link.

But do you find this file confusing? I would add more comments but I just don't know what to add there.

File: modules/API/API.js (hunk 25 above)
Comment (hristoterezov) [inline modules/API/API.js]: Thanks for the links. That's awesome.

Fixed!

File: modules/remotecontrol/Receiver.js (hunk 10 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: Let's do that with another PR. This is already big enough and it also started conflicting. WDYT?

File: modules/remotecontrol/Controller.js (hunk 19 above)
Comment (hristoterezov) [inline modules/remotecontrol/Controller.js]: Done!

File: modules/UI/videolayout/RemoteVideo.js (hunk 24 above)
Comment (hristoterezov) [inline modules/UI/videolayout/RemoteVideo.js]: I changed it to pin the controlled participant only if permissions have been granted (the remote control session start).

File: conference.js (hunk 21 above)
Comment (hristoterezov) [inline conference.js]: Done!

File: modules/remotecontrol/Receiver.js (hunk 10 above)
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: Actually Receiver class is kind of static for the whole app, so I can live with that.

File: modules/remotecontrol/Controller.js [hunk 27]
```diff
@@ -76,6 +77,7 @@ export default class Controller extends RemoteControlParticipant {
         if(!this.enabled) {
             return Promise.reject(new Error("Remote control is disabled!"));
         }
+        logger.debug("Requsting remote control permissions from: " + userId);
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Why this is logged on debug level ? Will this message appear much more often that the other added in this commit ? For me it would make more sense to have it logged on the "log" level as well.

File: modules/remotecontrol/Controller.js [hunk 28]
```diff
@@ -135,6 +137,8 @@ export default class Controller extends RemoteControlParticipant {
             switch(remoteControlEvent.action) {
                 case PERMISSIONS_ACTIONS.grant: {
                     this.controlledParticipant = userId;
+                    logger.debug("Remote control permissions granted to: "
```
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Same as the comment above, will it make sense to have it logged on the "log" level ?

File: modules/remotecontrol/Receiver.js [hunk 29]
```diff
@@ -128,6 +135,7 @@ export default class Receiver extends RemoteControlParticipant {
             APP.conference.addConferenceListener(ConferenceEvents.USER_LEFT,
                 this._userLeftListener);
             this.controller = userId;
+            logger.debug("Remote control permissions granted to: " + userId);
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: Same as the comment above, will it make sense to have it logged on the "log" level ?

File: modules/remotecontrol/RemoteControl.js (hunk 13 above)
Comment (paweldomas) [inline modules/remotecontrol/RemoteControl.js]: Is it possible then that the APP.API.init() will initialise twice ? I do not see any protection against that in the API module.

File: modules/remotecontrol/Receiver.js [hunk 30]
```diff
@@ -0,0 +1,191 @@
...
+        this._userLeftListener = this._onUserLeft.bind(this);
+        this._hangupListener = this._onHangup.bind(this);
+    }
+
+    /**
+     * Enables / Disables the remote control
+     * @param {boolean} enabled the new state.
+     */
+    enable(enabled) {
+        if(this.enabled !== enabled) {
```
Comment (paweldomas) [inline modules/remotecontrol/Receiver.js]: If this method is called twice with the same argument then it will execute twice. Is that ok ?

It can be at any point in future that something changes and make it execute twice (by mistake). I think we should detect that and either do not execute twice or log a warning.

File: modules/remotecontrol/Controller.js (hunk 23 above)
Comment (paweldomas) [inline modules/remotecontrol/Controller.js]: Maybe we don't, but it was never clear, so far...

File: modules/keycode/keycode.js (hunk 26 above)
Comment (paweldomas) [inline modules/keycode/keycode.js]: Just leaving a comment here that we've discussed it offline and this is Jitsi Meet API for the remote control. The value are the same as in RobotJS, but only for convenience.

File: modules/remotecontrol/RemoteControl.js (hunk 13 above)
Comment (hristoterezov) [inline modules/remotecontrol/RemoteControl.js]: We are checking if postis was initialized already: 
https://github.com/jitsi/jitsi-meet/pull/1192/files#diff-6e17c8c4f062496776d635d89187d13cR121
